src/iosource/BPF_Program.cc

```diff
@@ -120,12 +120,19 @@ bool BPF_Program::Compile(int snaplen, int linktype, const char* filter, uint32_
 	int err = pcap_compile_nopcap(snaplen, linktype, &m_program, (char*)filter, optimize, netmask,
 	                              my_error);
 	if ( err < 0 )
-		errbuf = std::string(my_error);
+		{
+		state = GetStateFromMessage(errstr);
+		state_message = util::fmt("pcap_compile(%s): %s", filter, pcap_geterr(pcap);
+		}
 #else
-	int err = pcap_compile_nopcap(snaplen, linktype, &m_program, (char*)filter, optimize, netmask);
+	int err = pcap_compile_nopcap(static_cast<int>(snaplen), linktype, &m_program, (char*)filter,
+	                              optimize, netmask);
 
+	// We have no way of knowing what the error actually was because pcap_compile_nocap doesn't
+	// return an error string nor any other information, so just assume every failure is
+	// fatal.
```

Comment:
It seems all that  `pcap_compile_nopcap()` does is to open a dead `pd` with the given snaplen and linktype and then call pcap_compile() on it.

https://github.com/the-tcpdump-group/libpcap/blob/d09102e61a868f96fb68dea8fcf08b06388bb450/gencode.c#L872-L885

As an alternative to the comment, we could probably duplicate those few lines and instantiate the pcap handle and then pass it on to
`BPF_Program::Compile(pcap_t* pcap, ...)` and capture the error there? It would also allow capturing/differentiate the `pcap_open_dead()` error.

Reply:
The manpage for pcap_compile_nopcap actually says this:

> A return of -1 indicates an error; the error text is unavailable. (pcap_compile_nopcap() is a wrapper around pcap_open_dead(), pcap_compile(), and pcap_close(); the latter three routines can be used directly in order to get the error text for a compilation error.) 

So yes, we could just duplicate that code and call `pcap_error` if one of those methods fails.

Reply:
Also, I think making that change would remove all of the concerns about the other version `pcap_compile_nopcap`, since we'd stop calling it anyways.

Reply:
> Also, I think making that change would remove all of the concerns about the other version pcap_compile_nopcap, since we'd stop calling it anyways.

Nice. Two birds one stone. The generic error makes sense to me, too.